File: sdks/typescript/src/wire/registry-proto-wire.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Manual protobuf wire encoding/decoding for plexspaces.object_registry.v1
// request/response messages used by WIT host imports from WASM.
// Matches sdks/go/plexspaces/registry_proto_wire.go.
//
// Field numbers from proto/plexspaces/v1/registry/object_registry.proto.

import { appendVarint, appendLengthDelimited, concatBytes, readVarint, skipField } from './proto-wire-common.js';

type WireU8 = Uint8Array<ArrayBuffer>;

const enc = new TextEncoder();
const dec = new TextDecoder();

function appendStringField(buf: WireU8, fieldNum: number, s: string): WireU8 {
  if (!s) return buf;
  const encoded = enc.encode(s);
  // Copy to an ArrayBuffer-backed Uint8Array to satisfy strict TS types.
  const bytes = new Uint8Array(encoded.length) as WireU8;
  bytes.set(encoded);
  const tag = fieldNum << 3 | 2;
  let b = appendVarint(buf, tag);
  b = appendVarint(b, bytes.length);
  return concatBytes(b, bytes);
}

function appendVarintField(buf: WireU8, fieldNum: number, v: number): WireU8 {
  const tag = fieldNum << 3; // wire type 0
  let b = appendVarint(buf, tag);
  return appendVarint(b, v);
}

// ObjectType enum values
export const ObjectType = {
  UNSPECIFIED: 0,
  ACTOR: 1,
  TUPLESPACE: 2,
  SERVICE: 3,
  VM: 4,
  APPLICATION: 5,
  WORKFLOW: 6,
  NODE: 7,
  PROCESS_GROUP: 8,
} as const;

export type ObjectTypeValue = typeof ObjectType[keyof typeof ObjectType];

function objectTypeToString(n: number): string {
  switch (n) {
    case 1: return 'actor';
    case 2: return 'tuplespace';
    case 3: return 'service';
    case 4: return 'vm';
    case 5: return 'application';
    case 6: return 'workflow';
    case 7: return 'node';
    case 8: return 'process_group';
    default: return '';
  }
}

function objectTypeFromString(s: string): number {
  switch (s) {
    case 'actor': return 1;
    case 'tuplespace': return 2;
    case 'service': return 3;
    case 'vm': return 4;
    case 'application': return 5;
    case 'workflow': return 6;
    case 'node': return 7;
    case 'process_group': return 8;
    default: return 0;
  }
}

export interface ObjectRegistration {
  objectId: string;
  objectType: string;
  grpcAddress?: string;
  objectCategory?: string;
  tenantId?: string;
  namespace?: string;
  capabilities?: string[];
  labels?: string[];
  alias?: string;
  healthStatus?: string;
}

// ============================================================================
// ObjectRegistration encoder
// Field numbers: object_id=1, object_type=3, tenant_id=5, namespace=6,
//                grpc_address=8, object_category=9, capabilities=10 (repeated),
//                labels=13 (repeated), alias=18
// ============================================================================

function encodeObjectRegistration(reg: ObjectRegistration): WireU8 {
  let b: WireU8 = new Uint8Array(0) as WireU8;
  b = appendStringField(b, 1, reg.objectId);
  const ot = objectTypeFromString(reg.objectType);
  if (ot !== 0) b = appendVarintField(b, 3, ot);
  if (reg.grpcAddress) b = appendStringField(b, 8, reg.grpcAddress);
  if (reg.objectCategory) b = appendStringField(b, 9, reg.objectCategory);
  if (reg.tenantId) b = appendStringField(b, 5, reg.tenantId);
  if (reg.namespace) b = appendStringField(b, 6, reg.namespace);
  for (const cap of (reg.capabilities ?? [])) b = appendStringField(b, 10, cap);
  for (const lbl of (reg.labels ?? [])) b = appendStringField(b, 13, lbl);
  if (reg.alias) b = appendStringField(b, 18, reg.alias);
  return b;
}

// ============================================================================
// RegisterRequest: registration=1 (message)
// ============================================================================

export function encodeRegisterRequest(reg: ObjectRegistration): WireU8 {
  const inner = encodeObjectRegistration(reg);
  return appendLengthDelimited(new Uint8Array(0) as WireU8, 1, inner);
}

// ============================================================================
// UnregisterRequest: object_id=1, object_type=2, tenant_id=3, namespace=4
// ============================================================================

export function encodeUnregisterRequest(objectId: string, objectType: number, tenantId?: string, namespace?: string): WireU8 {
  let b: WireU8 = new Uint8Array(0) as WireU8;
  b = appendStringField(b, 1, objectId);
  if (objectType !== 0) b = appendVarintField(b, 2, objectType);
  if (tenantId) b = appendStringField(b, 3, tenantId);
  if (namespace) b = appendStringField(b, 4, namespace);
  return b;
}

// ============================================================================
// LookupRequest: object_id=1, object_type=2, tenant_id=3, namespace=4, alias=5
// ============================================================================

export function encodeLookupRequest(objectId: string, objectType: number, tenantId?: string, namespace?: string, alias?: string): WireU8 {
  let b: WireU8 = new Uint8Array(0) as WireU8;
  if (objectId) b = appendStringField(b, 1, objectId);
  if (objectType !== 0) b = appendVarintField(b, 2, objectType);
  if (tenantId) b = appendStringField(b, 3, tenantId);
  if (namespace) b = appendStringField(b, 4, namespace);
  if (alias) b = appendStringField(b, 5, alias);
  return b;
}

// ============================================================================
// DiscoverRequest: object_type=1, object_category=2, tenant_id=4, namespace=5,
//                  capabilities=6, labels=7, page_size=10
// ============================================================================

export function encodeDiscoverRequest(opts: {
  objectType?: number;
  objectCategory?: string;
  tenantId?: string;
  namespace?: string;
  capabilities?: string[];
  labels?: string[];
  pageSize?: number;
}): WireU8 {
  let b: WireU8 = new Uint8Array(0) as WireU8;
  if (opts.objectType) b = appendVarintField(b, 1, opts.objectType);
  if (opts.objectCategory) b = appendStringField(b, 2, opts.objectCategory);
  if (opts.tenantId) b = appendStringField(b, 4, opts.tenantId);
  if (opts.namespace) b = appendStringField(b, 5, opts.namespace);
  for (const cap of (opts.capabilities ?? [])) b = appendStringField(b, 6, cap);
  for (const lbl of (opts.labels ?? [])) b = appendStringField(b, 7, lbl);
  if (opts.pageSize && opts.pageSize > 0) b = appendVarintField(b, 10, opts.pageSize);
  return b;
}

// ============================================================================
// HeartbeatRequest: object_id=1, object_type=2, tenant_id=3, namespace=4
// ============================================================================

export function encodeHeartbeatRequest(objectId: string, objectType: number, tenantId?: string, namespace?: string): WireU8 {
  let b: WireU8 = new Uint8Array(0) as WireU8;
  b = appendStringField(b, 1, objectId);
  if (objectType !== 0) b = appendVarintField(b, 2, objectType);
  if (tenantId) b = appendStringField(b, 3, tenantId);
  if (namespace) b = appendStringField(b, 4, namespace);
  return b;
}

// ============================================================================
// ObjectRegistration decoder
// ============================================================================

function decodeObjectRegistration(data: Uint8Array): ObjectRegistration {
  const reg: ObjectRegistration = { objectId: '', objectType: '' };
  let pos = 0;
  while (pos < data.length) {
    const { value: tagVal, n } = readVarint(data, pos);
    pos += n;
    const fn_ = Number(tagVal >> 3n);
    const wt = Number(tagVal & 7n);
    if (wt === 2) {
      const { value: ln, n: m } = readVarint(data, pos);
      pos += m;
      const end = pos + Number(ln);
      const chunk = data.slice(pos, end);
      pos = end;
      const str = dec.decode(chunk);
      switch (fn_) {
        case 1: reg.objectId = str; break;
        case 5: reg.tenantId = str; break;
        case 6: reg.namespace = str; break;
        case 8: reg.grpcAddress = str; break;
        case 9: reg.objectCategory = str; break;
        case 10: (reg.capabilities ??= []).push(str); break;
        case 13: (reg.labels ??= []).push(str); break;
        case 18: reg.alias = str; break;
      }
    } else if (wt === 0) {
      const { value: v, n: m } = readVarint(data, pos);
      pos += m;
      if (fn_ === 3) reg.objectType = objectTypeToString(Number(v));
    } else {
      pos = skipField(data, pos, wt);
    }
  }
  return reg;
}

// ============================================================================
// LookupResponse decoder: registration=1 (message), found=2 (bool)
// ============================================================================

export function decodeLookupResponse(data: Uint8Array): ObjectRegistration | null {
  let pos = 0;
  let regBytes: Uint8Array | null = null;
  let found = false;
  while (pos < data.length) {
    const { value: tagVal, n } = readVarint(data, pos);
    pos += n;
    const fn_ = Number(tagVal >> 3n);
    const wt = Number(tagVal & 7n);
    if (fn_ === 1 && wt === 2) {
      const { value: ln, n: m } = readVarint(data, pos);
      pos += m;
      regBytes = data.slice(pos, pos + Number(ln));
      pos += Number(ln);
    } else if (fn_ === 2 && wt === 0) {
      const { value: v, n: m } = readVarint(data, pos);
      pos += m;
      found = v !== 0n;
    } else {
      pos = skipField(data, pos, wt);
    }
  }
  if (!found || !regBytes) return null;
  return decodeObjectRegistration(regBytes);
}

// ============================================================================
// DiscoverResponse decoder: registrations=1 (repeated message)
// ============================================================================

export function decodeDiscoverResponse(data: Uint8Array): ObjectRegistration[] {
  const results: ObjectRegistration[] = [];
  let pos = 0;
  while (pos < data.length) {
    const { value: tagVal, n } = readVarint(data, pos);
    pos += n;
    const fn_ = Number(tagVal >> 3n);
    const wt = Number(tagVal & 7n);
    if (fn_ === 1 && wt === 2) {
      const { value: ln, n: m } = readVarint(data, pos);
      pos += m;
      const regBytes = data.slice(pos, pos + Number(ln));
      pos += Number(ln);
      results.push(decodeObjectRegistration(regBytes));
    } else {
      pos = skipField(data, pos, wt);
    }
  }
  return results;
}

File: sdks/go/plexspaces/registry_proto_wire.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// Protobuf wire encoding/decoding for plexspaces.object_registry.v1 request/response
// messages used by WASM host imports. Hand-coded to avoid importing generated proto
// packages in TinyGo WASM builds. Matches prost output in crates/wasm-runtime.
//
// Field numbers are taken directly from proto/plexspaces/v1/registry/object_registry.proto.

package plexspaces

// ============================================================================
// ObjectRegistration encoder
// Field numbers: object_id=1, object_name=2, object_type=3, tenant_id=5,
//                namespace=6, grpc_address=8, object_category=9,
//                capabilities=10 (repeated string), labels=13 (repeated string),
//                alias=18
// ============================================================================

func encodeObjectRegistration(reg ObjectRegistration) []byte {
	var b []byte
	b = appendStringField(b, 1, reg.ObjectID)
	if reg.ObjectType != "" {
		ot := objectTypeFromString(reg.ObjectType)
		if ot != 0 {
			b = appendVarintField(b, 3, uint64(ot))
		}
	}
	if reg.GRPCAddress != "" {
		b = appendStringField(b, 8, reg.GRPCAddress)
	}
	if reg.ObjectCategory != "" {
		b = appendStringField(b, 9, reg.ObjectCategory)
	}
	if reg.TenantID != "" {
		b = appendStringField(b, 5, reg.TenantID)
	}
	if reg.Namespace != "" {
		b = appendStringField(b, 6, reg.Namespace)
	}
	for _, cap := range reg.Capabilities {
		b = appendStringField(b, 10, cap)
	}
	for _, lbl := range reg.Labels {
		b = appendStringField(b, 13, lbl)
	}
	if reg.Alias != nil && *reg.Alias != "" {
		b = appendStringField(b, 18, *reg.Alias)
	}
	return b
}

// objectTypeFromString maps ObjectRegistration.ObjectType string to proto enum value.
func objectTypeFromString(s string) int32 {
	switch s {
	case "actor", "OBJECT_TYPE_ACTOR":
		return 1
	case "tuplespace", "OBJECT_TYPE_TUPLESPACE":
		return 2
	case "service", "OBJECT_TYPE_SERVICE":
		return 3
	case "vm", "OBJECT_TYPE_VM":
		return 4
	case "application", "OBJECT_TYPE_APPLICATION":
		return 5
	case "workflow", "OBJECT_TYPE_WORKFLOW":
		return 6
	case "node", "OBJECT_TYPE_NODE":
		return 7
	case "process_group", "OBJECT_TYPE_PROCESS_GROUP":
		return 8
	}
	return 0
}

func objectTypeToString(n int32) string {
	switch n {
	case 1:
		return "actor"
	case 2:
		return "tuplespace"
	case 3:
		return "service"
	case 4:
		return "vm"
	case 5:
		return "application"
	case 6:
		return "workflow"
	case 7:
		return "node"
	case 8:
		return "process_group"
	}
	return ""
}

// ============================================================================
// RegisterRequest: registration=1 (message)
// ============================================================================

func encodeRegisterRequest(reg ObjectRegistration) []byte {
	inner := encodeObjectRegistration(reg)
	return appendLengthDelimited(nil, 1, inner)
}

// ============================================================================
// UnregisterRequest: object_id=1, object_type=2, tenant_id=3, namespace=4
// ============================================================================

func encodeUnregisterRequest(objectID string, objectType int32, tenantID, namespace string) []byte {
	var b []byte
	b = appendStringField(b, 1, objectID)
	if objectType != 0 {
		b = appendVarintField(b, 2, uint64(objectType))
	}
	if tenantID != "" {
		b = appendStringField(b, 3, tenantID)
	}
	if namespace != "" {
		b = appendStringField(b, 4, namespace)
	}
	return b
}

// ============================================================================
// LookupRequest: object_id=1, object_type=2, tenant_id=3, namespace=4, alias=5
// ============================================================================

func encodeLookupRequest(objectID string, objectType int32, tenantID, namespace, alias string) []byte {
	var b []byte
	if objectID != "" {
		b = appendStringField(b, 1, objectID)
	}
	if objectType != 0 {
		b = appendVarintField(b, 2, uint64(objectType))
	}
	if tenantID != "" {
		b = appendStringField(b, 3, tenantID)
	}
	if namespace != "" {
		b = appendStringField(b, 4, namespace)
	}
	if alias != "" {
		b = appendStringField(b, 5, alias)
	}
	return b
}

// ============================================================================
// DiscoverRequest: object_type=1, object_category=2, tenant_id=4, namespace=5,
//                  capabilities=6, labels=7, page_size=10
// ============================================================================

func encodeDiscoverRequest(objectType int32, objectCategory, tenantID, namespace string,
	capabilities, labels []string, pageSize int32) []byte {
	var b []byte
	if objectType != 0 {
		b = appendVarintField(b, 1, uint64(objectType))
	}
	if objectCategory != "" {
		b = appendStringField(b, 2, objectCategory)
	}
	if tenantID != "" {
		b = appendStringField(b, 4, tenantID)
	}
	if namespace != "" {
		b = appendStringField(b, 5, namespace)
	}
	for _, cap := range capabilities {
		b = appendStringField(b, 6, cap)
	}
	for _, lbl := range labels {
		b = appendStringField(b, 7, lbl)
	}
	if pageSize > 0 {
		b = appendVarintField(b, 10, uint64(pageSize))
	}
	return b
}

// ============================================================================
// HeartbeatRequest: object_id=1, object_type=2, tenant_id=3, namespace=4
// ============================================================================

func encodeHeartbeatRequest(objectID string, objectType int32, tenantID, namespace string) []byte {
	var b []byte
	b = appendStringField(b, 1, objectID)
	if objectType != 0 {
		b = appendVarintField(b, 2, uint64(objectType))
	}
	if tenantID != "" {
		b = appendStringField(b, 3, tenantID)
	}
	if namespace != "" {
		b = appendStringField(b, 4, namespace)
	}
	return b
}

// ============================================================================
// LookupResponse decoder: registration=1 (message), found=2 (bool)
// ============================================================================

func decodeLookupResponse(data []byte) (*ObjectRegistration, bool) {
	pos := 0
	var regBytes []byte
	found := false
	for pos < len(data) {
		tag, n, err := readVarint(data, pos)
		if err != nil {
			break
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		switch {
		case fn == 1 && wt == 2: // registration message
			ln, m, err := readVarint(data, pos)
			if err != nil {
				break
			}
			pos += m
			end := pos + int(ln)
			if end > len(data) {
				break
			}
			regBytes = data[pos:end]
			pos = end
		case fn == 2 && wt == 0: // found bool
			v, m, err := readVarint(data, pos)
			if err != nil {
				break
			}
			pos += m
			found = v != 0
		default:
			var err error
			pos, err = skipField(data, pos, wt)
			if err != nil {
				break
			}
		}
	}
	if !found || len(regBytes) == 0 {
		return nil, false
	}
	reg := decodeObjectRegistration(regBytes)
	return &reg, true
}

// ============================================================================
// DiscoverResponse decoder: registrations=1 (repeated message)
// ============================================================================

func decodeDiscoverResponse(data []byte) []ObjectRegistration {
	var results []ObjectRegistration
	pos := 0
	for pos < len(data) {
		tag, n, err := readVarint(data, pos)
		if err != nil {
			break
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		if fn == 1 && wt == 2 {
			ln, m, err := readVarint(data, pos)
			if err != nil {
				break
			}
			pos += m
			end := pos + int(ln)
			if end > len(data) {
				break
			}
			reg := decodeObjectRegistration(data[pos:end])
			results = append(results, reg)
			pos = end
		} else {
			var err error
			pos, err = skipField(data, pos, wt)
			if err != nil {
				break
			}
		}
	}
	return results
}

// ============================================================================
// ObjectRegistration decoder
// ============================================================================

func decodeObjectRegistration(data []byte) ObjectRegistration {
	var reg ObjectRegistration
	pos := 0
	for pos < len(data) {
		tag, n, err := readVarint(data, pos)
		if err != nil {
			break
		}
		pos += n
		fn := tag >> 3
		wt := tag & 7
		switch {
		case wt == 2: // length-delimited (string, bytes, embedded message)
			ln, m, err := readVarint(data, pos)
			if err != nil {
				break
			}
			pos += m
			end := pos + int(ln)
			if end > len(data) {
				break
			}
			chunk := string(data[pos:end])
			pos = end
			switch fn {
			case 1:
				reg.ObjectID = chunk
			case 2:
				// object_name - ignore
			case 5:
				reg.TenantID = chunk
			case 6:
				reg.Namespace = chunk
			case 8:
				reg.GRPCAddress = chunk
			case 9:
				reg.ObjectCategory = chunk
			case 10:
				reg.Capabilities = append(reg.Capabilities, chunk)
			case 13:
				reg.Labels = append(reg.Labels, chunk)
			case 18:
				alias := chunk
				reg.Alias = &alias
			}
		case wt == 0: // varint
			v, m, err := readVarint(data, pos)
			if err != nil {
				break
			}
			pos += m
			switch fn {
			case 3: // object_type enum
				reg.ObjectType = objectTypeToString(int32(v))
			case 12: // health_status enum
				// ignore for now
			}
		default:
			var err error
			pos, err = skipField(data, pos, wt)
			if err != nil {
				break
			}
		}
	}
	return reg
}

// ============================================================================
// Low-level proto encoding helpers (wire format)
// appendVarint/appendLengthDelimited defined in tuplespace_proto_wire.go
// ============================================================================

func appendStringField(buf []byte, fieldNum int, s string) []byte {
	tag := uint64(fieldNum<<3 | 2) // wire type 2 = length-delimited
	buf = appendVarint(buf, tag)
	buf = appendVarint(buf, uint64(len(s)))
	return append(buf, s...)
}

func appendVarintField(buf []byte, fieldNum int, v uint64) []byte {
	tag := uint64(fieldNum << 3) // wire type 0 = varint
	buf = appendVarint(buf, tag)
	return appendVarint(buf, v)
}
